File: ios/Wrappers/DmWrapper.swift
//
//  DmWrapper.swift
//  Pods
//
//  Created by Naomi Plasterer on 10/24/24.
//

import Foundation
import XMTP

// Wrapper around XMTP.Dm to allow passing these objects back into react native.
struct DmWrapper {
	static func encodeToObj(_ dm: XMTP.Dm, client: XMTP.Client, conversationParams: ConversationParamsWrapper = ConversationParamsWrapper()) async throws -> [String: Any] {
		var result: [String: Any] = [
			"clientInboxId": client.inboxID,
			"id": dm.id,
			"createdAt": UInt64(dm.createdAt.timeIntervalSince1970 * 1000),
			"version": "DM",
			"topic": dm.topic,
			"peerInboxId": try dm.peerInboxId,
			"commitLogForkStatus": ConversationDebugInfoWrapper.commitLogForkStatusToString(dm.commitLogForkStatus())
		]

		if conversationParams.consentState {
			result["consentState"] = ConsentWrapper.consentStateToString(state: try dm.consentState())
		}
		if conversationParams.lastMessage {
			if let lastMessage = try await dm.lastMessage() {
				result["lastMessage"] = try MessageWrapper.encode(lastMessage)
			}
		}
		
		return result
	}

	static func encode(_ dm: XMTP.Dm, client: XMTP.Client, conversationParams: ConversationParamsWrapper = ConversationParamsWrapper()) async throws -> String {
		let obj = try await encodeToObj(dm, client: client, conversationParams: conversationParams)
		let data = try JSONSerialization.data(withJSONObject: obj)
		guard let result = String(data: data, encoding: .utf8) else {
			throw WrapperError.encodeError("could not encode dm")
		}
		return result
	}
}

File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/DmWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.GsonBuilder
import org.xmtp.android.library.Client
import org.xmtp.android.library.Dm

class DmWrapper {
    companion object {
        suspend fun encodeToObj(
            client: Client,
            dm: Dm,
            dmParams: ConversationParamsWrapper = ConversationParamsWrapper(),
        ): Map<String, Any> {
            return buildMap {
                put("clientInboxId", client.inboxId)
                put("id", dm.id)
                put("createdAt", dm.createdAt.time)
                put("version", "DM")
                put("topic", dm.topic)
                put("peerInboxId", dm.peerInboxId)
                put("commitLogForkStatus", ConversationDebugInfoWrapper.commitLogForkStatusToString(dm.commitLogForkStatus()))
                if (dmParams.consentState) {
                    put("consentState", consentStateToString(dm.consentState()))
                }
                if (dmParams.lastMessage) {
                    val lastMessage = dm.lastMessage()
                    if (lastMessage != null) {
                        put("lastMessage", MessageWrapper.encode(lastMessage))
                    }
                }
            }
        }

        suspend fun encode(
            client: Client,
            dm: Dm,
            dmParams: ConversationParamsWrapper = ConversationParamsWrapper(),
        ): String {
            val gson = GsonBuilder().create()
            val obj = encodeToObj(client, dm, dmParams)
            return gson.toJson(obj)
        }
    }
}
